File: rosduino/handlers.cpp
/*
 * This is a place for ros subscribers and publishers
 * Please keep larger handler functions outside this file
 */
#include "handlers.h"
#include "odrive.h"

/*
 * SUBSCRIBERS
 */

// example

//void send_back_cb( const std_msgs::String& cmd_msg){
//  back_msg = cmd_msg;
//  back_sub.publish( &back_msg );
//  digitalWrite(13, HIGH-digitalRead(13));  //toggle led  
//}
//
//ros::Subscriber<std_msgs::String> send_back_sub("send_back", send_back_cb);

void _position_cb( const std_msgs::Float32& distance){
  odrive_run_fixed(0, distance.data);

//      for (float ph = 0.0f; ph < 6.28318530718f; ph += 0.01f) {
//        float pos_m0 = 20000.0f * cos(ph);
////        float pos_m1 = 20000.0f * sin(ph);
//        odrives[0]->SetPosition(0, pos_m0);
////        odrive.SetPosition(1, pos_m1);
//        delay(5);
//      }
  digitalWrite(13, HIGH-digitalRead(13));  //toggle led
}

ros::Subscriber<std_msgs::Float32> position_sub("position", _position_cb);
//
//void _drive_limits_cb( const std_msgs::Float32MultiArray& limits){
//  odrive_set_limits(0, limits.data[0], limits.data[1]);
//  digitalWrite(13, HIGH-digitalRead(13));  //toggle led
//}
//
//ros::Subscriber<std_msgs::Float32MultiArray> drive_limits_sub("drive/limits", _drive_limits_cb);

void _drive_commands_cb( const std_msgs::Char& command){
  if(command.data == 'r') odrive_init();
  digitalWrite(13, HIGH-digitalRead(13));  //toggle led
}

ros::Subscriber<std_msgs::Char> drive_commands_sub("drive/commands", _drive_commands_cb);
//
//void _drive_position_cb( const std_msgs::Float32MultiArray& positions){
//  odrive_drive_fixed(positions.data[0], positions.data[1]);
//  digitalWrite(13, HIGH-digitalRead(13));  //toggle led
//}
//
//ros::Subscriber<std_msgs::Float32MultiArray> drive_position_sub("drive/position", _drive_position_cb);

/*
 * PUBLISHERS
 */

//std_msgs::String rosout_msg;
//ros::Publisher rosout_pub("rosout", &rosout_msg);

/*
 * FUNCTIONS
 */

void handlers_init() {
//  nh.subscribe(send_back_sub); // example
//  nh.advertise(rosout_pub);
  nh.subscribe(position_sub);
//  nh.subscribe(drive_limits_sub);
//  nh.subscribe(drive_position_sub);
  nh.subscribe(drive_commands_sub);
//  nh.subscribe(reset_sub);
  
  
  
}

File: catkin_ws/src/nuc_main/src/handlers.py
import rospy  # in PyCharm shows error, solution: https://gavazzi.us/2017/07/31/pycharm-ros/
from std_msgs import msg

import drive
from config import *

# SUBSCRIBER CALLBACKS

def _position_test_cb(position):
    drive.set_axis_position(drive.odrives[ODRIVE_DRIVE_ID].axis0, position.data)
    rospy.loginfo("Recieved on position_test: " + str(position.data))


def _drive_commands_cb(command):
    if chr(command.data) == 'r':
        drive.full_reset_and_calibrate_all()
    rospy.loginfo("Recieved on drive/commands: " + str(command.data))


def _drive_distance_cb(distances):
    drive.drive_distance(distances.data[0], distances.data[1])
    rospy.loginfo("Recieved on drive/distance: " + ' '.join(map(str, distances.data)))


def _drive_velocity_cb(velocities):
    drive.drive_velocity(velocities.data[0], velocities.data[1])
    rospy.loginfo("Recieved on drive/velocity: " + ' '.join(map(str, velocities.data)))


def _drive_vel_limits_cb(velocities):
    drive.set_vel_limits(velocities.data)
    rospy.loginfo("Recieved on drive/velocity_limits: " + ' '.join(map(str, velocities.data)))


def _drive_acc_limits_cb(accs):
    drive.set_acc_limits(accs.data)
    rospy.loginfo("Recieved on drive/velocity_limits: " + ' '.join(map(str, accs.data)))

# PUBLISHERS

def debug_publish(msg_str):
    try:
        debug_pub.publish(msg_str)
        rospy.loginfo("Sent on debug: " + str(msg_str.data))
    except rospy.ROSInterruptException:
        pass

def axis_states_publish(axes):
    try:
        msg_arr = msg.Float32MultiArray()
        msg_arr.data = axes
        msg_arr.layout.dim = [msg.MultiArrayDimension(), msg.MultiArrayDimension(), msg.MultiArrayDimension()]
        msg_arr.layout.dim[0].size = len(axes)
        msg_arr.layout.dim[1].size = len(axes[0])
        msg_arr.layout.dim[2].size = len(axes[0][0])
        msg_arr.layout.dim[0].stride = len(axes) * len(axes[0]) * len(axes[0][0])
        msg_arr.layout.dim[1].stride = len(axes[0]) * len(axes[0][0])
        msg_arr.layout.dim[2].stride = len(axes[0][0])
        msg_arr.layout.dim[0].label = "odrives"
        msg_arr.layout.dim[1].label = "axes"
        msg_arr.layout.dim[2].label = "values"

        print(msg_arr.data)

        print(msg_arr)

        axis_state_pub.publish(msg_arr)
        # rospy.loginfo("Sent on debug: " + str(axes))
    except rospy.ROSInterruptException:
        pass

# TEST

def _send_back(msg_str):
    debug_publish(msg_str.data)

# FUNCTIONS

def init_handlers():
    rospy.init_node('nuc_main', anonymous=True)

    # SUBSCRIBERS

    rospy.Subscriber("position_test", msg.Float32, _position_test_cb)
    rospy.Subscriber("drive/commands", msg.Char, _drive_commands_cb)
    rospy.Subscriber("drive/distance", msg.Float32MultiArray, _drive_distance_cb)
    rospy.Subscriber("drive/velocity", msg.Float32MultiArray, _drive_velocity_cb)
    rospy.Subscriber("drive/vel_limits", msg.Float32MultiArray, _drive_vel_limits_cb)
    rospy.Subscriber("drive/acc_limits", msg.Float32MultiArray, _drive_acc_limits_cb)

    rospy.Subscriber("send_back", msg.String, _send_back) # TEST

    global debug_pub
    debug_pub = rospy.Publisher('debug', msg.String, queue_size=10)

    global axis_state_pub
    axis_state_pub = rospy.Publisher('drive/axis_states', msg.Float32MultiArray, queue_size=10)


# TODO ?
#   void _drive_limits_cb( const std_msgs::Float32MultiArray& limits){
#     odrive_set_limits(0, limits.data[0], limits.data[1]);
#     digitalWrite(13, HIGH-digitalRead(13));  //toggle led
#   }
#   ros::Subscriber<std_msgs::Float32MultiArray> drive_limits_sub("drive/limits", _drive_limits_cb);
